source/agora/test/Byzantine.d

```diff
@@ -304,29 +294,54 @@ unittest
     txes.each!(tx => node_1.putTransaction(tx));
     network.setTimeFor(Height(1));  // trigger consensus
 
-    waitForCount(conf.validators - 1, &network.nominate_count, "nominate");
-    waitForCount(conf.validators - 1, &network.prepare_count, "prepare");
-    waitForCount(conf.validators - 1, &network.confirm_count, "confirm");
-    waitForCount(conf.validators - 1, &network.externalize_count, "externalize");
+    waitForCount(conf.validators - 1, &network.envelope_type_counts.nominate_count, "nominate");
+    waitForCount(conf.validators - 1, &network.envelope_type_counts.prepare_count, "prepare");
+    waitForCount(conf.validators - 1, &network.envelope_type_counts.confirm_count, "confirm");
+    waitForCount(conf.validators - 1, &network.envelope_type_counts.externalize_count, "externalize");
 }
 
-/// Too many byzantine nodes will prevent block from being externalized
+
+// This delay is unfortunately needed before checking the number of confirmed envelopes
+private enum Duration sufficient_time_for_first_envelope = 500.msecs;
+
+/// Half nodes not signing correctly WILL NOT prevent block from being externalized if we require 3 out of 6
 unittest
 {
-    TestConf conf = { validators : 10, quorum_threshold : 66 };
+    TestConf conf = { validators : 6, quorum_threshold : 50 };
     auto network = makeTestNetwork!(ByzantineManager!(true, 1, 1))(conf);
     network.start();
     scope(exit) network.shutdown();
     scope(failure) network.printLogs();
     network.waitForDiscovery();
-
     auto nodes = network.clients;
     auto node_1 = nodes[$ - 1];
+    assert(node_1.getQuorumConfig().threshold == 3); // We should need 3 nodes
     auto txes = genesisSpendable().map!(txb => txb.sign()).array();
     txes.each!(tx => node_1.putTransaction(tx));
     network.setTimeFor(Height(1));  // trigger consensus
+    waitForCount(conf.validators - 1, &network.envelope_type_counts.nominate_count, "nominate");
+    Thread.sleep(sufficient_time_for_first_envelope);
```

Comment:
Why not just use `waitForCount` here?

Reply:
This is used to check that the time we wait in the next test is sufficient. As we want to avoid polling for up to 5 secs we just need to know our choice of delay is sufficient. It is not ideal and hopefully can be improved otherwise it may become another source of flakey testing.

Reply:
There should be a `network.expectBlock` call in here somewhere. Then after that call you can check all these message counts, without any delays.

Reply:
But the expectBlock waits for a period of time also.

Reply:
The counting of messaging is trying to avoid that polling with delay for the negative test case.

Reply:
> But the expectBlock waits for a period of time also.

In the future we can replace the implementation so it doesn't poll. And if all the code uses that function then we only need to fix that one function.

I think this hardcoded `Thread.sleep` call will fail arbitrarily in the test-suite based on the system load.

And finally, `expectBlock` actually tests the expected output of an Agora node. I think it's very important that we test the output of the node, and not focus on testing SCP internals.

Reply:
Following our discussion I have put back the waitForBlock with up to 3 second waits on the positive tests except for the last which confirms the delay is long enough for the negative test where we expect no confirm envelopes to be received. The negative test can someday be replaced with a better scenario in which a later round of consensus is successful but without the Byzantine nodes being more than the allowed threshold.